File: php/bank-account/BankAccount.php
<?php declare(strict_types=1);

class BankAccount
{
    public function __construct(public int $balance = 0, private bool $open = false) {}
    public function open()
    {
        if ($this->open)
            throw new Exception("account already open");
        $this->open = true;
        $this->balance = 0;
    }

    public function close()
    {
        if (!$this->open)
            throw new Exception("account not open");
        $this->open = false;
    }

    public function balance(): int
    {
        if (!$this->open)
            throw new Exception("account not open");

        return $this->balance;
    }

    public function deposit(int $amt)
    {
        if (!$this->open)
            throw new Exception("account not open");
        if ($amt < 0)
            throw new InvalidArgumentException("amount must be greater than 0");

        $this->balance += $amt;
    }

    public function withdraw(int $amt)
    {
        if (!$this->open)
            throw new Exception("account not open");
                if ($amt < 0)
            throw new InvalidArgumentException("amount must be greater than 0");

        if ($this->balance < $amt)
            throw new InvalidArgumentException("amount must be less than balance");

        $this->balance -= $amt;
    }
}

File: java/bank-account/src/main/java/BankAccount.java
class BankAccount {
    private boolean opened = false;
    private int balance = 0;

    void open() throws BankAccountActionInvalidException {
        synchronized(this) {
            if (opened) {
                throw new BankAccountActionInvalidException("Account already open");
            }

            this.balance = 0;
            this.opened = true;
        }
    }

    void close() throws BankAccountActionInvalidException {
        synchronized(this) {
            if (!opened) {
                throw new BankAccountActionInvalidException("Account not open");
            }

            this.opened = false;
        }
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        synchronized(this) {
            if (!opened) {
                throw new BankAccountActionInvalidException("Account closed");
            }

            return this.balance;
        }
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        synchronized(this) {
            if (!opened) {
                throw new BankAccountActionInvalidException("Account closed");
            }

            if (amount < 0) {
                throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
            }

            this.balance += amount;
        }
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        synchronized(this) {
            if (!opened) {
                throw new BankAccountActionInvalidException("Account closed");
            }

            if (amount < 0) {
                throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
            }

            if (amount > this.balance) {
                throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
            }

            this.balance -= amount;
        }
    }

}